File: tests/helpers/produtos.py
from helpers.auth import headers

def criar_produto(client, token):

    response = client.post(
        "/produtos",
        headers=headers(token),
        json={
            "nome": "Produto Teste",
            "valor": 10.5,
            "estoque": 5
        }
    )

    return response.json()["dados"]

def criar_produto_com_nome_valor_estoque(client, token, nome, valor, estoque):

    response = client.post(
        "/produtos",
        headers=headers(token),
        json={
            "nome": nome,
            "valor": valor,
            "estoque": estoque
        }
    )

    return response

def buscar_produto_por_id(client, token, produto_id):
    
    response = client.get(
        f"/produtos/{produto_id}",
        headers=headers(token)
    )

    return response.json()["dados"]

def buscar_produto_por_nome(client, token, nome):

    response = client.get(
        f"/produtos?nome={nome}",
        headers=headers(token)
    )

    return response.json()["dados"]

def listar_produtos_ativos(client, token):

    response = client.get(
        "/produtos",
        headers=headers(token)
    )

    return response.json()["dados"]

def atualizar_produto(client, token, nome, valor, estoque):

    produto = criar_produto(client, token)

    response = client.put(
        f"/produtos/{produto['id']}",
        headers=headers(token),
        json={
            "nome": nome,
            "valor": valor,
            "estoque": estoque
        }
    )

    return response

def inativar_produto(client, token, produto_id):

    response = client.put(
    f"/produtos/{produto_id}/inativar",
    headers=headers(token)
    )

    return response.json()["dados"]

def reativar_produto(client, token, produto_id):

    response = client.put(
        f"/produtos/{produto_id}/reativar",
        headers=headers(token)
    )

    return response.json()["dados"]

def listar_produtos_inativos(client, token):

    response = client.get(
        "/produtos/inativos",
        headers=headers(token)
    )

    return response.json()["dados"]

File: frontend/js/produtos.js
const token = localStorage.getItem("token");

if (!token) {
    mostrarMensagem("Faça login para acessar", "erro");
    window.location.href = "login.html";
}

function ir(pagina) {
    window.location.href = pagina;
}

function logout() {
    localStorage.removeItem("token");
    window.location.href = "login.html";
}

function renderizarProdutos(data) {
    const lista = document.getElementById("listaProdutos");
    lista.innerHTML = "";

    data.dados.forEach(p => {
        const tr = document.createElement("tr");

        tr.innerHTML = `
            <td>${p.nome}</td>
            <td>R$ ${p.valor.toFixed(2)}</td>
            <td>${p.estoque}</td>
            <td>
                <button onclick="editarProduto(
                ${p.id},
                '${p.nome}',
                ${p.valor},
                ${p.estoque})">Editar</button>
                <button  class="btn-danger" 
                onclick="inativarProduto(${p.id})">Inativar</button>
                `;

        lista.appendChild(tr);
    });
}

function editarProduto(id, nome, valor, estoque) {

    document.getElementById("nomeProduto").value = nome;
    document.getElementById("preco").value = valor;
    document.getElementById("estoque").value = estoque

    window.produtoEditando = id;

    document.getElementById("btnSalvar").innerText = "Atualizar";
}

function cancelarEdicao() {
    window.produtoEditando = null;

    document.getElementById("nomeProduto").value = "";
    document.getElementById("preco").value = "";
    document.getElementById("estoque").value = "";
    document.getElementById("btnSalvar").innerText = "Salvar";

}

async function criarProduto() {
    const nome = document.getElementById("nomeProduto").value.trim();
    const preco = parseFloat(document.getElementById("preco").value);
    const estoque = parseInt(document.getElementById("estoque").value);

    if (!nome || isNaN(preco) || isNaN(estoque)) {
        mostrarMensagem("Preencha todos os campos corretamente", "erro");
        return;
    }


    let endpoint = "/produtos/";
    let method = "POST";

    if (window.produtoEditando) {
        endpoint += window.produtoEditando;
        method = "PUT";
    }

    try {

        await api(endpoint, {
            method: method,
            body: JSON.stringify({
                nome: nome,
                valor: preco,
                estoque: estoque
            })
        });

        mostrarMensagem(window.produtoEditando ? "Produto atualizado!" : "Produto criado!", "sucesso");

        window.produtoEditando = null;

        document.getElementById("nomeProduto").value = "";
        document.getElementById("preco").value = "";
        document.getElementById("estoque").value = "";
        document.getElementById("btnSalvar").innerText = "Salvar";

        listarProdutos();

    } catch (error) {
        mostrarMensagem("Erro ao salvar produto", "erro");
        console.error(error);

        listarProdutos();
    }
}

async function listarProdutos() {

    const data = await api("/produtos");
    renderizarProdutos(data);
}

async function buscarProduto() {
    const nome = document.getElementById("buscarProduto").value.trim();

    const data = await api(`/produtos?nome=${nome}`);
    renderizarProdutos(data);
}

async function inativarProduto(id) {
    const confirmar = confirm("tem certeza que deseja inativar ? Você poderá reativar novamente ");
    if (!confirmar) return;

    try {
        await api(`/produtos/${id}/inativar`, {
            method: "PUT"
        });

        mostrarMensagem("Produto inativado com sucesso!", "sucesso");

        listarProdutos();

    } catch (error) {
        mostrarMensagem("Erro ao inativar produto", "erro");
        console.error(error);
    }

}